x-pack/plugins/apm/server/lib/services/get_service_transaction_groups_agg_results.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  EVENT_OUTCOME,
+  SERVICE_NAME,
+  TRANSACTION_DURATION,
+  TRANSACTION_NAME,
+  TRANSACTION_TYPE,
+} from '../../../common/elasticsearch_fieldnames';
+import { EventOutcome } from '../../../common/event_outcome';
+import { LatencyAggregationType } from '../../../common/latency_aggregation_types';
+import { rangeFilter } from '../../../common/utils/range_filter';
+import { Coordinate } from '../../../typings/timeseries';
+import {
+  getDocumentTypeFilterForAggregatedTransactions,
+  getProcessorEventForAggregatedTransactions,
+  getTransactionDurationFieldForAggregatedTransactions,
+} from '../helpers/aggregated_transactions';
+import { getBucketSize } from '../helpers/get_bucket_size';
+import {
+  getLatencyAggregation,
+  getLatencyValue,
+} from '../helpers/latency_aggregation_type';
+import { Setup, SetupTimeRange } from '../helpers/setup_request';
+
+export async function getServiceTransactionGroupsAggResults({
+  serviceName,
+  transactionNames,
+  setup,
+  numBuckets,
+  searchAggregatedTransactions,
+  transactionType,
+  latencyAggregationType,
+}: {
+  serviceName: string;
+  transactionNames: string[];
+  setup: Setup & SetupTimeRange;
+  numBuckets: number;
+  searchAggregatedTransactions: boolean;
+  transactionType: string;
+  latencyAggregationType: LatencyAggregationType;
+}): Promise<
+  Record<
+    string,
+    {
+      latency: Coordinate[];
+      throughput: Coordinate[];
+      errorRate: Coordinate[];
+      impact: number;
+    }
+  >
+> {
+  const { apmEventClient, start, end, esFilter } = setup;
+  const deltaAsMinutes = (end - start) / 1000 / 60;
+
+  const { intervalString } = getBucketSize({ start, end, numBuckets });
+
+  const field = getTransactionDurationFieldForAggregatedTransactions(
+    searchAggregatedTransactions
+  );
+
+  const response = await apmEventClient.search({
+    apm: {
+      events: [
+        getProcessorEventForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      ],
+    },
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { [SERVICE_NAME]: serviceName } },
+            { term: { [TRANSACTION_TYPE]: transactionType } },
+            { range: rangeFilter(start, end) },
+            ...getDocumentTypeFilterForAggregatedTransactions(
+              searchAggregatedTransactions
+            ),
+            ...esFilter,
+          ],
+        },
+      },
+      aggs: {
+        total_duration: { sum: { field: TRANSACTION_DURATION } },
+        transaction_name_filter: {
+          filter: { terms: { [TRANSACTION_NAME]: transactionNames } },
```

Comment:
instead of a filter agg, can you try a terms agg with a filter (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_filtering_values_4) and compare the performance?

Reply:
@dgieselaar FWICT using the terms agg with `include` prop is better than the filter agg.

Filter agg runs one TermQuery for each filtered value:
<img width="1364" alt="Screenshot 2021-02-03 at 11 26 14" src="https://user-images.githubusercontent.com/55978943/106734040-ddb4f300-6612-11eb-9b4e-d39437170bcf.png">

While the terms agg does not:
<img width="1379" alt="Screenshot 2021-02-03 at 11 28 54" src="https://user-images.githubusercontent.com/55978943/106734197-089f4700-6613-11eb-83cf-d4c8e4131327.png">

Reply:
FWICT you're looking at query performance, not aggregation performance, and you're looking at span data, not transaction data?

Reply:
Ups, my bad. These are the aggregation results.

With filter agg:
<img width="1366" alt="Screenshot 2021-02-03 at 13 30 56" src="https://user-images.githubusercontent.com/55978943/106747429-3640bc00-6624-11eb-9a38-0616d3948b0c.png">

With term agg with include prop:
<img width="1375" alt="Screenshot 2021-02-03 at 13 31 26" src="https://user-images.githubusercontent.com/55978943/106747462-40fb5100-6624-11eb-9e18-d724a4f6d505.png">